### PR TITLE
[Demangler] Further optimizations for the remangler.

### PR DESCRIPTION
**Pull request by al45tair**
Replace use of `snprintf()` with some custom code (this is around 10x faster on my machine).

Move a few of the `Node` member functions to the header so they're inlined.

Optimize the `deepEquals()` function by adding an `isSimilarTo()` method on `Node`; the checks that were happening in the `deepEquals()` function could be implemented more efficiently by making use of details of the internal representation of `Node`.

rdar://125739630

### CODE REVIEW

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
This is a further 23% reduction in cost for the mangling I'm presently testing, over and above what I've already managed.

**Comment by al45tair** [comment]
@swift-ci Please smoke test

File: test/Demangle/Inputs/bigtype-objcrt.txt [hunk 1]
```diff
@@ -1 +1 @@
...
...t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_t_tN
```
Comment (mikeash) [inline test/Demangle/Inputs/bigtype-objcrt.txt:1]: Is the loss of positional info expected or is "pos" some sort of WIP placeholder?

Comment (al45tair) [reply test/Demangle/Inputs/bigtype-objcrt.txt:1]: It's a placeholder.

**Review by mikeash — APPROVED** [review]
I'm informed that `pos` is expected because the exact line/column changes when you change the remangler code which is a real pain to deal with and not useful to test.

**Comment by tbkka** [comment]
Do these need to be cherry-picked to 6.0?

**Pull request merged**